collections/users.js

```diff
@@ -178,12 +179,13 @@ Meteor.methods({
 		// causes us to fail.
 
 		var user = Meteor.user();
+		if (!user) return ApiError("plzLogin", "Not logged-in");
 
 		const saneUsername = saneText(username).trim().substring(0, 200);
 		if (saneUsername && user.username !== saneUsername) {
 			let result = Profile.Username.change(user._id, saneUsername, "profile change");
 			if (!result) {
-				throw new Meteor.Error("nameError", "Failed to update username");
+				return ApiError("nameError", "Failed to update username");
```

Comment:
The idea of latency compensation is to run a method on both, the server and the client and just hope that the result is the same. If it is not, then the server wins, which leads to the result on the client being updated. There is a blog post on [discover meteor](https://www.discovermeteor.com/blog/advanced-latency-compensation/) with details and numerous examples. The important bit is the following: The client version of the method is executed **before** a request to execute the method is sent to the server.

The aforementioned post also discusses the role of the callback (when a method is invoked on the client). It is important to realize that this callback is only invoked **after** the server returned its result. Hence even if the stub method (the one invoked on the client) already detected invalid input, the user is only notified after the server roundtrip(!). It follows that the throw-on-server-return-on-client-pattern introduced here (also explained in [that blog post](http://smashingthingstogether.com/handling-errors-in-meteor-methods/)) indeed eliminates the logged Exception but it does not help with latency compensation in any way. The returned exception is simply swallowed by `Meteor.call` / `Meteor.apply`. We still need to wait for the server to throw before being able to notify the user.

That raises the question why to bother with latency compensation at all. Why not just define the methods on the server and forget about the stub on the client. In fact grepping through the meteor core packages suggests that there is no single stub-method definition (i.e. client side method):

```
~/src/meteor/packages(devel)$ git grep -li meteor.method | grep -v test
accounts-base/accounts_server.js
accounts-password/password_server.js
check/README.md
ddp-rate-limiter/README.md
ddp-server/livedata_server.js
dynamic-import/server.js
meteor-base/README.md
non-core/bundle-visualizer/server.js
```

Hence, maybe it would be better to just define the methods in server code and forget about stub-methods in general. We still can add them in special cases and with special logic in order to improve user experience. There is also a [gits from arunoda](https://gist.github.com/arunoda/5552469) which suggests that approach.